Added IDA* implementation

tsa87 opened the pull request:
Implemented Iterative Deepening A* Search in `search.py`

tsa87 committed:
Added IDA* implementation

tsa87 commented:
First open-source contribution :)
Please advise revisions.

antmarakis commented on search.py:
```diff
@@ -420,8 +420,35 @@ def astar_search(problem, h=None, display=False):
     return best_first_graph_search(problem, lambda n: n.path_cost + h(n), display)
 
 
+def cost_limited_astar_search(problem, limit, f):
```
You need to comment the code a bit, unfortunately right now it may be a bit hard to follow.

tsa87 replied:
Hey, thanks for the feedback! I added more comments and descriptions for the search function. Please feel free to let me know if they are not clear.

antmarakis commented on search.py:
```diff
@@ -420,8 +420,35 @@ def astar_search(problem, h=None, display=False):
+def cost_limited_astar_search(problem, limit, f):
+
+    def recursive_clas(node, problem, limit, f):
+        if problem.goal_test(node.state):
```
For example, add some description for each if statement

antmarakis commented on search.py:
```diff
@@ -420,8 +420,35 @@ def astar_search(problem, h=None, display=False):
 
+def cost_limited_astar_search(problem, limit, f):
+
+    def recursive_clas(node, problem, limit, f):
```
Is that a type? What is 'clas'?

tsa87 replied:
I changed the function name to "recursive_cost_limited_astar_search"

antmarakis commented on search.py:
```diff
@@ -420,8 +420,35 @@ def astar_search(problem, h=None, display=False):
+            return 'cutoff'
+        else:
+            cutoff_occurred = False
+            for child in node.expand(problem):
```
Here it would also be nice to have a quick comment explaining what is going on

tsa87 committed:
Merge https://github.com/aimacode/aima-python

tsa87 committed:
Added comments

dmeoli committed:
Merge remote-tracking branch 'origin/master' into pr1175-work

dmeoli committed:
Implement IDA* (iterative-deepening A*) correctly

Rewrite iterative_deepening_astar_search to use f = g + h contours: a
depth-first search pruned at f(node) > bound, raising the bound to the
smallest f that exceeded it, with path-based cycle avoidance. The previous
version pruned on h alone, decremented a depth counter, had a cutoff_occured
typo and no cycle handling (it looped on graphs). Add a test against A*.
